### ISSUE TITLE
Tests of org.goobi.production.flow.statistics.StatisticsManagerTest failing randomly

### ISSUE THREAD

**Issue by henning-gerhardt** [goobi/goobi-production#424]
Running test-suite with `ant test` many times, it will happen that tests inside `org.goobi.production.flow.statistics.StatisticsManagerTest` are failing:

```
junit:
    [junit] Running de.sub.goobi.helper.FilesystemHelperTest
    [junit] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.189 sec
    [junit] Running de.sub.goobi.helper.archive.ProcessSwapOutTaskTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.005 sec
    [junit] Running de.sub.goobi.helper.encryption.DesEncrypterTest
    [junit] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.057 sec
    [junit] Running de.sub.goobi.helper.encryption.MD4Test
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec
    [junit] Running de.sub.goobi.helper.encryption.MD5Test
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 sec
    [junit] Running de.sub.goobi.helper.importer.ImportOpacTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.004 sec
    [junit] Running de.sub.goobi.metadaten.PaginatorTest
    [junit] Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.079 sec
    [junit] Running de.sub.goobi.samples.BenutzerTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.003 sec
    [junit] Running de.sub.goobi.samples.BenutzergruppenTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.004 sec
    [junit] Running org.goobi.io.BackupFileRotationTest
    [junit] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.045 sec
    [junit] Running org.goobi.production.GoobiVersionTest
    [junit] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 sec
    [junit] Running org.goobi.production.flow.statistics.StatisticsManagerTest
    [junit] Tests run: 25, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 0.142 sec
    [junit] TEST org.goobi.production.flow.statistics.StatisticsManagerTest FAILED
    [junit] Running org.goobi.production.flow.statistics.StatisticsRenderingElementTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.003 sec
    [junit] Running org.goobi.production.flow.statistics.enums.CalculationUnitTest
    [junit] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
    [junit] Running org.goobi.production.flow.statistics.enums.ResultOutputTest
    [junit] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
    [junit] Running org.goobi.production.flow.statistics.enums.StatisticsModeTest
    [junit] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec
    [junit] Running org.goobi.production.flow.statistics.enums.TimeUnitTest
    [junit] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.ConverterTest
    [junit] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.SQLHelperTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.SQLProductionTest
    [junit] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.SQLStepRequestsTest
    [junit] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.SQLStorageTest
    [junit] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.StatQuestCorrectionsTest
    [junit] Tests run: 6, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.009 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.StatQuestProductionTest
    [junit] Tests run: 6, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.014 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.StatQuestProjectAssociationsTest
    [junit] Tests run: 5, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.008 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.StatQuestStorageTest
    [junit] Tests run: 6, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.01 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.StatQuestThroughputTest
    [junit] Tests run: 6, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.008 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.UserDefinedFilterTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.003 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.UserProcessesFilterTest
    [junit] Tests run: 10, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 0.022 sec
    [junit] Running org.goobi.production.flow.statistics.hibernate.UserTemplatesFilterTest
    [junit] Tests run: 10, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 0.013 sec
    [junit] Running org.goobi.webapi.beans.IdentifierPPNTest
    [junit] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 sec
    [junit] Tests FAILED
```

Running test for this class through `ant run-test -Djunit.test.name=org.goobi.production.flow.statistics.StatisticsManagerTest` results after many runs in no failure.:

```
run-test:
    [junit] Testsuite: org.goobi.production.flow.statistics.StatisticsManagerTest
    [junit] Tests run: 25, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.238 sec
    [junit] 
    [junit] Testcase: testStatisticsManager took 0.006 sec
    [junit] Testcase: testGetJfreeDataset took 0.007 sec
    [junit] Testcase: testGetStatisticMode took 0.004 sec
    [junit] Testcase: testCalculate took 0 sec
    [junit] 	SKIPPED: Crashing
    [junit] Testcase: testGetAllTimeUnits took 0.004 sec
    [junit] Testcase: testGetAllCalculationUnits took 0.005 sec
    [junit] Testcase: testGetAllResultOutputs took 0.004 sec
    [junit] Testcase: testGetSourceDateFrom took 0.004 sec
    [junit] Testcase: testSetSourceDateFrom took 0.003 sec
    [junit] Testcase: testGetSourceDateTo took 0.003 sec
    [junit] Testcase: testSetSourceDateTo took 0.003 sec
    [junit] Testcase: testGetSourceNumberOfTimeUnitsAsString took 0.003 sec
    [junit] Testcase: testSetSourceNumberOfTimeUnitsAsString took 0.002 sec
    [junit] Testcase: testGetSourceTimeUnit took 0.003 sec
    [junit] Testcase: testSetSourceTimeUnit took 0.003 sec
    [junit] Testcase: testGetTargetTimeUnit took 0.003 sec
    [junit] Testcase: testSetTargetTimeUnit took 0.002 sec
    [junit] Testcase: testGetTargetResultOutput took 0.002 sec
    [junit] Testcase: testSetTargetResultOutput took 0.002 sec
    [junit] Testcase: testGetTargetCalculationUnit took 0.003 sec
    [junit] Testcase: testSetTargetCalculationUnit took 0.003 sec
    [junit] Testcase: testIsShowAverage took 0.003 sec
    [junit] Testcase: testSetShowAverage took 0.003 sec
    [junit] Testcase: testGetRenderingElements took 0.005 sec
    [junit] Testcase: testGetLocale took 0.003 sec
```

affected version: >= 1.10.x (and maybe 1.8.x)
system: Debian 7 / 8
ant version: provided by Debian 7 / 8

**Comment by henning-gerhardt** [comment]
With my pull-request #440 I was able to get some more information from `ant test` in `build/test-reports/TEST-org.goobi.production.~StatisticsManagerTest.xml` file:

```xml
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceDateFrom" time="0.008">
    <failure message="expected:&lt;Sun Feb 01 00:00:00 CET 2009&gt; but was:&lt;Mon Feb 02 00:00:00 CET 2009&gt;" type="junit.framework.AssertionFailedError">junit.framework.AssertionFailedError: expected:&lt;Sun Feb 01 00:00:00 CET 2009&gt; but was:&lt;Mon Feb 02 00:00:00 CET 2009&gt;
	at org.goobi.production.flow.statistics.StatisticsManagerTest.testGetSourceDateFrom(StatisticsManagerTest.java:168)
</failure>
  </testcase>

```

**Comment by henning-gerhardt** [comment]
Don't read my previous comment not wrong: this information is available without pull-request #440 too but with this change retrieving this information is much easier.

**Comment by henning-gerhardt** [comment]
Ok, I think I root case of this issue: tests are not executed how they are declared, they are executed in parallel or random order. If existing two tests where on get a value and one other set a value then it is possible that set value test are executed before get test and result of this is that get test failed.

Logged test execution show this. First failed test suite
```xml
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceDateFrom" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceDateTo" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceDateTo" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceNumberOfTimeUnitsAsString" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceNumberOfTimeUnitsAsString" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetResultOutput" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetResultOutput" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetCalculationUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetCalculationUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testIsShowAverage" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetShowAverage" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetRenderingElements" time="0.005" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetLocale" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testStatisticsManager" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetJfreeDataset" time="0.006" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetStatisticMode" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testCalculate" time="0.0">
    <skipped message="Crashing" />
  </testcase>
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllTimeUnits" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllCalculationUnits" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllResultOutputs" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceDateFrom" time="0.008">
    <failure message="expected:&lt;Sun Feb 01 00:00:00 CET 2009&gt; but was:&lt;Mon Feb 02 00:00:00 CET 2009&gt;" type="junit.framework.AssertionFailedError">junit.framework.AssertionFailedError: expected:&lt;Sun Feb 01 00:00:00 CET 2009&gt; but was:&lt;Mon Feb 02 00:00:00 CET 2009&gt;
	at org.goobi.production.flow.statistics.StatisticsManagerTest.testGetSourceDateFrom(StatisticsManagerTest.java:168)
</failure>
  </testcase>

```

second successful run of same test suite

```xml
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testStatisticsManager" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetJfreeDataset" time="0.006" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetStatisticMode" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testCalculate" time="0.0">
    <skipped message="Crashing" />
  </testcase>
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllTimeUnits" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllCalculationUnits" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetAllResultOutputs" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceDateFrom" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceDateFrom" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceDateTo" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceDateTo" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceNumberOfTimeUnitsAsString" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceNumberOfTimeUnitsAsString" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetSourceTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetSourceTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetTimeUnit" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetResultOutput" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetResultOutput" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetTargetCalculationUnit" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetTargetCalculationUnit" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testIsShowAverage" time="0.002" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testSetShowAverage" time="0.003" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetRenderingElements" time="0.006" />
  <testcase classname="org.goobi.production.flow.statistics.StatisticsManagerTest" name="testGetLocale" time="0.003" />
```

Difference is execution of tests `testSetSourceDateFrom` and `testGetSourceDateFrom`. If `testSetSourceDateFrom` is executed before `testGetSourceDateFrom` then test suite are failing. In any other cases test suite are successful.
